### PR TITLE
Proper command-line parsing

### PR DESCRIPTION
**Pull request by mehrdadn**
## Why are these changes needed?

We currently split command-lines on spaces.

## Related issue number

#631

## Checks

- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://ray.readthedocs.io/en/latest/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failure rates at https://ray-travis-tracker.herokuapp.com/.
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested (please justify below)

### CODE REVIEW

File: src/ray/util/process.cc [hunk 1]
```diff
@@ -54,7 +55,15 @@ class ProcessFD {
...
-    fd = _spawnvp(P_NOWAIT, argv[0], argv);
+    std::vector<std::string> args;
+    for (size_t i = 0; argv[i]; ++i) {
+      args.push_back(argv[i]);
+    }
+    // Calling CreateCommandLine() here wouldn't make sense here if the
+    // Microsoft C runtime properly quoted each command-argument argument.
+    // However, it doesn't quote at all. It just joins arguments with a space.
+    // So we have to do the quoting manually and pass everything as a single argument.
+    fd = _spawnlp(P_NOWAIT, args[0].c_str(), CreateCommandLine(args).c_str(), NULL);
```
Comment (pcmoritz) [inline src/ray/util/process.cc:66]: Calling `.c_str()` on the temporary string returned from `CreateCommandLine(args)` is undefined behaviour (and can blow up), you should store the result in a temporary variable.

Comment (mehrdadn) [reply src/ray/util/process.cc:66]: It's not undefined behavior :) objects are only destroyed at the end of the statement, at which point `_spawnlp` has already returned.

Comment (pcmoritz) [reply src/ray/util/process.cc:66]: looks like it is defined after all: https://stackoverflow.com/a/22146708

File: src/ray/util/util_test.cc [hunk 2]
```diff
@@ -0,0 +1,129 @@
...
+#else
+      proc = syn == posix ? popen(test_command.c_str(), "r") : NULL;
+#endif
+      if (proc) {
+        ASSERT_TRUE(fgets(&*buf.begin(), static_cast<int>(buf.size()), proc));
+        buf = std::string(buf.c_str());  // remove trailing nulls
+        ASSERT_TRUE(!buf.empty());
+        ASSERT_EQ(buf.back(), '\n');
+        buf.pop_back();
+        ASSERT_EQ(ParseCommandLine(cmdline, syn), arglist);
```
Comment (pcmoritz) [inline src/ray/util/util_test.cc:107]: We already tested this above in `ASSERT_EQ(ParseCommandLine(CreateCommandLine(arglist, syn), syn), arglist);`, didn't we? Is this line supposed to test something involving `buf`?

File: src/ray/util/util.h [hunk 3]
```diff
@@ -94,6 +101,12 @@ struct EnumClassHash {
   }
 };
 
+/// Uses sscanf() to read a token matching from the string, advancing the iterator.
+/// \param c_str A string iterator that is dereferenceable. (i.e.: c_str < string::end())
+/// \param format The pattern. It must not produce any output. (e.g., use %*d, not %d.)
+/// \return The scanned prefix of the string, if any.
+std::string ScanToken(std::string::const_iterator &c_str, std::string format);
```
Comment (pcmoritz) [inline src/ray/util/util.h:108]: Let's make this function private. We also have a copy of it in `url.cc`, so maybe we can unify them (e.g. by putting everything into `util.cc`).

File: src/ray/util/util_test.cc (hunk 2 above)
Comment (mehrdadn) [reply src/ray/util/util_test.cc:107]: Oops, yeah! Good eye, I'll fix it. Hopefully this hasn't let bugs slip through!

File: src/ray/util/util.h (hunk 3 above)
Comment (mehrdadn) [reply src/ray/util/util.h:108]: Okay sure.

**Review by pcmoritz — APPROVED** [review]

**Pull request merged**